Explain networking (#299)
FabricMC/fabric-docs: develop/networking.md

## Why Is Networking Important? {#why-is-networking-important}

> Packets are the core concept of networking in Minecraft. Packets are made up of arbitrary data that can be sent either from
> server to client or from client to server. Almost everything you do in the game involves networking in some way, whether
> you know it or not. For example, when you send a chat message, a packet is sent to the server with the content. The server then
> sends another packet to all the other clients with your message.

Reviewer: Either this:

```suggestion
Packets are the core concept of networking in Minecraft.
Packets are made up of arbitrary data that can be sent either from server to client or from client to server.
Check out the diagram below, which provides a visual representation of the networking architecture in Fabric:

![Sided Architecture](/assets/develop/networking/sides.png)

Notice how packets are the bridge between the server and the client; that's because almost everything you do in the game involves networking in some way, whether you know it or not.
For example, when you send a chat message, a packet is sent to the server with the content.
The server then sends another packet to all the other clients with your message.
```

...or delete [sides.png](https://github.com/FabricMC/fabric-docs/blob/daff2047d7cb7ebcb4c12e37164622449ce0c871/public/assets/develop/networking/sides.png) if deemed unnecessary.

<details>
<summary>Sided Architecture</summary>

![Sided Architecture](https://github.com/FabricMC/fabric-docs/blob/daff2047d7cb7ebcb4c12e37164622449ce0c871/public/assets/develop/networking/sides.png?raw=true)

</details>

> One imporant thing to keep in mind is there is always a server running, even in singleplayer and LAN. Packets are still used to communicate between

Reviewer 2: ```suggestion
One important thing to keep in mind is there is always a server running, even in singleplayer and LAN. Packets are still used to communicate between
```

## Why Is Networking Important? {#why-is-networking-important}

> Packets of data are the core concept of networking in Minecraft. Packets are made up of arbitrary data that can be sent either from

Reviewer 2: ```suggestion
Packets are the core concept of networking in Minecraft. Packets are made up of arbitrary data that can be sent either from
```

> the server to the client or from the client to the server. Almost everything you do in the game involves networking in some way, whether

Reviewer 2: ```suggestion
server to client or from client to server. Almost everything you do in the game involves networking in some way, whether
```

you know it or not. For example, when you jump in the game, a packet is sent to the server to tell it that you jumped. The server then

> sends a packet to all the other clients to tell them that you jumped.

Reviewer 2: I'm not sure how much it matters, but this doesn't sound factually correct 😅 AFAIK a jump action (and other forms of player movement) just gets boiled down to a change in position and velocity

Author: Thats true, ill change this to a chat message. Just wanted something simple to get the idea across.

> One imporant thing to keep in mind is even in single player there is still a server running, Packets are still used to communicate between

Reviewer 2: ```suggestion
One imporant thing to keep in mind is there is always a server running, even in singleplayer and LAN. Packets are still used to communicate between
```

> the client and server even when no one else is playing with you. We usually call this the logicial client/server or the client/server thread.
> This is diffrent from the client/server code split, that prevents you from being able to use client only classes in common (shared between
> the client and server) code.

Reviewer 2: ```suggestion
the client and server even when no one else is playing with you. When talking about sides in networking, the terms "**logical** client" and "**logical** server" are used. The integrated singleplayer/LAN server and the dedicated server are both **logical** servers, but only the dedicated server can be considered a **physical** server.
```

> When state is not synced between the client and server, you can run into issues where the server or other clients dont agree with what the

Reviewer 2: ```suggestion
When state is not synced between the client and server, you can run into issues where the server or other clients don't agree with what another
```

client is doing. This is often known as a "desync". When writing your own mod you may need to send a packet of data to notify the server,

> client or any other connected clients that something has happened.

Reviewer 2: It is the server's job to notify other clients about changes. Merging "notify server (from client)" and "notify client (from server)" in the same sentence like this is confusing

Author: Hopefully its better now.
